### PR TITLE
feat: Proxy client

### PR DESCRIPTION
**Pull request by epociask**
Closes:
- https://github.com/Layr-Labs/eigenda-proxy/issues/32 
- https://github.com/Layr-Labs/eigenda-proxy/issues/30

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+// ProxyClient is an interface for communicating with the EigenDA proxy server
+type ProxyClient interface {
+	Health() error
+	GetData(ctx context.Context, cert *common.Certificate, domain common.DomainType) ([]byte, error)
+	SetData(ctx context.Context, b []byte) (*common.Certificate, error)
+}
+
+// client is the implementation of ProxyClient
+type client struct {
```
Comment (teddyknox) [inline client/client.go:33]: * Why not expose the concrete type?
* Naming-wise, I think "eigenda" should either be in the package name or in the struct/interface name.

File: client/client.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	cfg        *Config
+	httpClient *http.Client
+}
+
+func New(cfg *Config) ProxyClient {
```
Comment (teddyknox) [inline client/client.go:38]: i.e. why not return a `client` type? (may need to make it a public type)

File: client/client.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
```
Comment (teddyknox) [inline client/client.go:59]: Would be nice to add a protocol for including error messages from the response body in the error output here.

File: client/client.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+// TODO: Add support for custom http client option
+type Config struct {
+	URL string
+}
+
+// ProxyClient is an interface for communicating with the EigenDA proxy server
+type ProxyClient interface {
+	Health() error
```
Comment (teddyknox) [inline client/client.go:27]: Is this part of the OP plasma interface? I figured this endpoint was for docker healthchecks.

File: client/client.go (hunk 1 above)
Comment (teddyknox) [reply client/client.go:33]: Is this client specific to OP stack or general?

File: client/client.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+// GetData fetches blob data associated with a DA certificate
+func (c *client) GetData(ctx context.Context, cert *common.Certificate, domain common.DomainType) ([]byte, error) {
+	b, err := rlp.EncodeToBytes(cert)
+	if err != nil {
+		return nil, err
+	}
+
+	// encode prefix bytes
+	b = eigenda.GenericPrefix(eigenda.Commitment(b).Encode())
```
Comment (teddyknox) [inline client/client.go:74]: nit: not terribly ergonomic to have `Commitment(b)` do nothing to `b` without a corresponding call to `.Encode()`. What if it looked more like

```
type genericPrefix String
type commitment String
b = eigenda.NewGenericPrefix(eigenda.NewCommitment(b))
```

where these are functions and not type casting operators and b

File: client/client.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to store data: %v", resp.StatusCode)
+	}
+
+	b, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(b) < decodingOffset {
```
Comment (teddyknox) [inline client/client.go:120]: nit: might rename this to something more obvious like "minAllowedCommitmentLength" or something like that

File: client/client.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+		return nil, fmt.Errorf("read certificate is of invalid length: %d", len(b))
+	}
+
+	var blob *common.Certificate
```
Comment (teddyknox) [inline client/client.go:124]: nit: rename `blob` var to `cert`

File: common/common.go [hunk 8]
```diff
@@ -93,3 +110,26 @@ func ParseBytesAmount(s string) (uint64, error) {
 		return 0, fmt.Errorf("unsupported unit: %s", unit)
 	}
 }
+
+func SubsetExists[P comparable](arr, sub []P) bool {
```
Comment (teddyknox) [inline common/common.go:114]: where is this being used?

File: client/client.go (hunk 5 above)
Comment (teddyknox) [reply client/client.go:74]: Update:

I'm now seeing that `GenericPrefix()` is already a function, but `Commitment()` is not. So my suggestion would be to make `Commitment()` into a simple function so the `Encode()` part can be dropped.

File: operator-setup
Comment (teddyknox) [inline operator-setup:1]: what does this submodule update accomplish? point it to my merged test srs points?

File: test/e2e_test.go [hunk 9]
```diff
@@ -138,7 +142,59 @@ func TestE2EPutGetLogicForEigenDAStore(t *testing.T) {
...
+	t.Log("Waiting for client to establish connection with plasma server...")
+	// wait for server to come online after starting
+	wait.For(ts.ctx, time.Second*1, func() (bool, error) {
+		err := daClient.Health()
+		if err != nil {
+			return false, nil
+		}
+
+		return true, nil
+	})
```
Comment (teddyknox) [inline test/e2e_test.go:166]: I believe that createTestSuite() is already waiting for the server to come up.

https://github.com/Layr-Labs/eigenda-proxy/blob/f270ca10fe3c7be238db3e71ef7d367f6454eeb5/server/server.go#L114

File: test/e2e_test.go [hunk 10]
```diff
@@ -165,3 +221,55 @@ func TestE2EPutGetLogicForMemoryStore(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, testPreimage, preimage)
 }
+
+func TestMemStoreWithProxyClient(t *testing.T) {
```
Comment (teddyknox) [inline test/e2e_test.go:225]: nit: Theoretically it should be possible to dry up these tests. Not necessary to do manually but maybe chatgpt can do.

**Review by teddyknox — APPROVED** [review]
So the idea here is that we have different versions of the proxy conform to the *effective* plasma interface of different versions of optimism, but then we maintain this client to hold a certain idealized interface constant through those changes. And this client can be used in Orbit, ZK Sync, CDK. Do I have that right?

Approving with comments.

File: operator-setup
Comment (epociask) [reply operator-setup:1]: precisely

File: test/e2e_test.go (hunk 9 above)
Comment (epociask) [reply test/e2e_test.go:166]: this just does a time wait and verifies that no error is generated within 10 ms. Feel like doing the await gurantees client <-> server communication is ready versus instantiation the server construct

File: client/client.go (hunk 1 above)
Comment (epociask) [reply client/client.go:33]: general

**Pull request merged**